src/gluonnlp/embedding/embed_loader.py

```diff
@@ -0,0 +1,238 @@
+                if word in vocab:
+                    index = vocab[word]
+                    matrix[index] = np.fromstring(' '.join(nums), sep=' ', dtype='float32', count=dim)
+                    hit_flags[index] = True
+            except Exception as e:
+                logging.error("Error occurred at the {} line.".format(idx))
+                raise e
+    return matrix, hit_flags
+
+def _load_embedding_npz(file_path, vocab, unknown):
+    hit_flags = np.zeros(len(vocab), dtype=bool)
+    npz_dict = np.load(file_path, allow_pickle=True)
+    unknown_token = npz_dict['unknown_token']
+    if not unknown_token:
+        unknown_token = None
+    else:
+        if isinstance(unknown_token, np.ndarray):
+            if unknown_token.dtype.kind == 'S':
+                unknown_token = unknown_token.tobytes().decode()
+            else:
+                unknown_token = str(unknown_token)
+    if unknown != unknown_token:
+        warnings.warn("You may not assign correct unknown token in the pretrained file"
+                      "Use {} as the unknown mark.".format(unknown_token))
+
+    idx_to_token = npz_dict['idx_to_token'].tolist()
+    token2idx = {x : i for i, x in enumerate(idx_to_token)}
+    idx_to_vec = nd.array(npz_dict['idx_to_vec'])
+    matrix = np.random.randn(len(vocab), idx_to_vec.shape[-1]).astype('float32')
+    for token, i in vocab.token_to_idx.items():
+        if token == vocab.unk_token and unknown_token is not None:
+            word = unknown_token
+        else:
+            word = token
+        if word in token2idx:
+            index = token2idx[word]
+            matrix[i] = idx_to_vec[index].asnumpy()
+            hit_flags[i] = True
+    return matrix, hit_flags
+
+def _get_file_url(cls_name, file_name):
+    namespace = 'gluon/embeddings/{}'.format(cls_name)
+    return _get_repo_file_url(namespace, file_name)
+
+def _check_and_get_path(pretrained_name_or_dir):
+    if os.path.exists(pretrained_name_or_dir):
+        return pretrained_name_or_dir
+    root_path = os.path.expanduser(os.path.join(get_home_dir(), 'embedding'))
+    for cls_name, embedding_cls in text_embedding_reg.items():
+        if pretrained_name_or_dir in embedding_cls:
+            source = pretrained_name_or_dir
+            embedding_dir = os.path.join(root_path, cls_name)
+            file_name, file_hash = embedding_cls[source]
+            url = _get_file_url(cls_name, file_name)
+            file_path = os.path.join(embedding_dir, file_name)
+            if not os.path.exists(file_path) or not check_sha1(file_path, file_hash):
+                logging.info('Embedding file {} is not found. Downloading from Gluon Repository. '
+                             'This may take some time.'.format(file_name))
+                download(url, file_path, sha1_hash=file_hash)
+            return file_path
+
+    return None
+
+def load_embeddings(vocab, pretrained_name_or_dir='glove.6B.50d', unknown='<unk>',
```

Comment:
How about to use separate classes for handling different embedding vectors. You may just implement some basic functions. For example, implement the `model.load()` functionality to load the parameter. And `model.similarity(a, b, method=None)` function to evaluate the similarity. Thus, it will be more structured and close to how we will evaluate it in the paper.

```python
class KeyedVector:
   def __init__(self):

   def load(cls, path):
      ...

   def similarity(self, a, b):
      ...
```

```python
class FastText
   def __init__(self):
      ...
   def load(cls, path):
      ...
```

Reply:
If we just load the embedding vectors from a small set of vocabulary, we can also add some flags to the load function, e.g., `def load(cls, path, vocab=None, num_tokens=None):`, in which `vocab` means a potential vocabulary object to help reduce the number of tokens to load. For example, the data augmentation algorithm in TinyBERT only loads the first 100000 words for fast lookup: https://github.com/huawei-noah/Pretrained-Language-Model/blob/e670706c041246b975a3646bc6a27c48786f6c15/TinyBERT/data_augmentation.py#L75

Reply:
I think we can set a base class `Class EmbeddingModel(HybridBlock)` to serve as the base class of embedding models.  And we can attach some evaluation functions to this class. For just loading a embedding matrix, we can simply use the current `load_embeddings` for users to manually set_data, or have `Class WordEmbedding(EmbeddingModel)` and move the functionality of `load_embeddings` to this class. And complex embedding models like `FastText`, character-level CNN can be further implemented based on `EmbeddingModel`. This embedding models may be implemented in `models/`